### ISSUE TITLE
KeyError: 'dealOrders'

### ISSUE THREAD

**Issue by slix88** [ccxt/ccxt#2212]
Hi

I am running a diagnostic on the Kucoin exchange and upon running the following code, I am getting a new error described as "KeyError: 'dealOrders'".

I am basically running the following code:
orders_kucoin = kucoin.fetch_closed_orders(symbol='NEO/BTC')
print(orders_kucoin)

...and I am getting the following error message:
Traceback (most recent call last):
  File "C:/.../CCXT_Binance_Kucoin_playground.py", line 197, in <module>
    orders_kucoin = kucoin.fetch_closed_orders(symbol='NEO/BTC')
  File "C:\...\lib\site-packages\ccxt\kucoin.py", line 504, in fetch_closed_orders
    }), market)
  File "C:\...\lib\site-packages\ccxt\kucoin.py", line 345, in parse_order
    trades = self.safe_value(order['dealOrders'], 'datas')
KeyError: 'dealOrders'

Could you please advise what is causing this error? I didn't receive this previously and it's only occurred since I did a recent pull of your code containing the latest updates.

Many thanks!

**Comment by kroitor** [comment]
Hi!

Can you please show your verbose output?

```
kucoin.verbose = True  # ←-------------- add this line
orders_kucoin = kucoin.fetch_closed_orders(symbol='NEO/BTC')
print(orders_kucoin)
```

Just copypaste the printout here (without your keys, of course). Thx!

**Comment by slix88** [comment]
Hi Kroitor,

Many thanks for the quick reply! I have produced an output for POWR as an example (fewer transactions)

`('\nRequest:', 'GET', 'https://api.kucoin.com/v1/order/dealt?limit=20&symbol=POWR-BTC', {'KC-API-SIGNATURE': 'XXX', 'KC-API-KEY': 'XXX', 'KC-API-NONCE': 'XXX', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36'}, None)
('\nResponse:', 'GET', 'https://api.kucoin.com/v1/order/dealt?limit=20&symbol=POWR-BTC', '200', "{'Access-Control-Allow-Methods': 'POST, GET, PUT, OPTIONS, DELETE, PATCH', 'KC-NODE': 'localhost', 'Access-Control-Max-Age': '86400', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Access-Control-Allow-Credentials': 'true', 'Date': 'Sun, 11 Mar 2018 01:47:50 GMT', 'Access-Control-Allow-Origin': 'https://www.kucoin.com', 'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept, Accept-Language', 'Content-Type': 'application/json;charset=UTF-8'}", u'{"success":true,"code":"OK","msg":"Operation succeeded.","timestamp":1520732870648,"data":{"total":1,"firstPage":true,"lastPage":false,"datas":[{"coinType":"POWR","createdAt":1517787481000,"amount":154.056735,"dealValue":0.01167596,"fee":0.15405674,"dealDirection":"BUY","coinTypePair":"BTC","oid":"5a779959410d9011ea625206","dealPrice":7.579E-5,"orderOid":"5a779958410d9014b6c07822","feeRate":0.001,"direction":"BUY"}],"currPageNo":1,"limit":20,"pageNos":1}}')
Traceback (most recent call last):
  File "C:/.../CCXT_Binance_Kucoin_playground.py", line 198, in <module>
    orders_kucoin = kucoin.fetch_closed_orders(symbol='POWR/BTC')
  File "C:\...\lib\site-packages\ccxt\kucoin.py", line 504, in fetch_closed_orders
    }), market)
  File "C:\...\lib\site-packages\ccxt\kucoin.py", line 345, in parse_order
    trades = self.safe_value(order['dealOrders'], 'datas')
KeyError: 'dealOrders'

Process finished with exit code 1`

**Comment by slix88** [comment]
Thanks for updating!

I was originally experiencing the issue in the python code. Could you please also commit to the kucoin.py code as well please?

**Comment by kroitor** [comment]
@slix88 it is auto-generated from JS as described here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#multilanguage-support

It will arrive in version 1.11.69 in 5-10 minutes. You just need to update your version of ccxt and it should work from there. We will be happy if you report back whether it works normally for you or not. Thx!

**Comment by kroitor** [comment]
The build has arrived just about now )

**Comment by slix88** [comment]
Apologies!

I have a couple more queries and wasn't sure if it needed to be posted as a "new issue". Both are in respect of **kucoin.py**.

First one is that "**fetch_closed_orders"** is no longer producing an error (which is the fix above). However, it's not returning anything. I have included an example below:

_Running code:_
```
    orders_kucoin = kucoin.fetch_closed_orders(symbol='POWR/BTC')
    print(orders_kucoin)
```

_Produces:_
```
('\nRequest:', 'GET', 'https://api.kucoin.com/v1/order/dealt?limit=20&symbol=POWR-BTC', {'KC-API-SIGNATURE': 'xxx', 'KC-API-KEY': 'xxx', 'KC-API-NONCE': 'xxx', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36'}, None)
('\nResponse:', 'GET', 'https://api.kucoin.com/v1/order/dealt?limit=20&symbol=POWR-BTC', '200', "{'Access-Control-Allow-Methods': 'POST, GET, PUT, OPTIONS, DELETE, PATCH', 'KC-NODE': 'kc-kitchen-ap3', 'Access-Control-Max-Age': '86400', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Access-Control-Allow-Credentials': 'true', 'Date': 'Sun, 11 Mar 2018 12:12:52 GMT', 'Access-Control-Allow-Origin': 'https://www.kucoin.com', 'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept, Accept-Language', 'Content-Type': 'application/json;charset=UTF-8'}", u'{"success":true,"code":"OK","msg":"Operation succeeded.","timestamp":1520770372867,"data":{"total":1,"firstPage":true,"lastPage":false,"datas":[{"coinType":"POWR","createdAt":1517787481000,"amount":154.056735,"dealValue":0.01167596,"fee":0.15405674,"dealDirection":"BUY","coinTypePair":"BTC","oid":"5a779959410d9011ea625206","dealPrice":7.579E-5,"orderOid":"5a779958410d9014b6c07822","feeRate":0.001,"direction":"BUY"}],"currPageNo":1,"limit":20,"pageNos":1}}')
[]
```
Note: The "[]" in the last line is producing a blank array

Second issue is that **"fetch_order"** only works for buy orders, not sell orders. Again, I have included an example below:

_Running code:_
```
    orders_kucoin = kucoin.fetch_closed_orders(symbol='XRB/BTC')
    print(orders_kucoin)
    buy_order_ID_kucoin = '5aa493dd0b936965a81c0d73' # NOTE: this is the order0id
    sell_order_ID_kucoin = '5aa499c50b9369664b72606e' # NOTE: this is the order0id
    last_buy_order_kucoin = kucoin.fetch_order(id=buy_order_ID_kucoin, symbol='XRB/BTC', params={'type': "BUY"})
    print(last_buy_order_kucoin)
    last_sell_order_kucoin = kucoin.fetch_order(id=sell_order_ID_kucoin, symbol='XRB/BTC', params={'type': "SELL"})
    print(last_sell_order_kucoin)
```

_Produces:_
```
('\nRequest:', 'GET', 'https://api.kucoin.com/v1/order/dealt?limit=20&symbol=XRB-BTC', {'KC-API-SIGNATURE': 'xxx', 'KC-API-KEY': 'xxx', 'KC-API-NONCE': 'xxx', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36'}, None)
('\nResponse:', 'GET', 'https://api.kucoin.com/v1/order/dealt?limit=20&symbol=XRB-BTC', '200', "{'Access-Control-Allow-Methods': 'POST, GET, PUT, OPTIONS, DELETE, PATCH', 'KC-NODE': 'localhost', 'Access-Control-Max-Age': '86400', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Access-Control-Allow-Credentials': 'true', 'Date': 'Sun, 11 Mar 2018 12:19:41 GMT', 'Access-Control-Allow-Origin': 'https://www.kucoin.com', 'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept, Accept-Language', 'Content-Type': 'application/json;charset=UTF-8'}", u'{"success":true,"code":"OK","msg":"Operation succeeded.","timestamp":1520770781039,"data":{"total":192,"firstPage":true,"lastPage":false,"datas":[{"coinType":"XRB","createdAt":1520736710000,"amount":0.95,"dealValue":0.00119319,"fee":1.19E-6,"dealDirection":"SELL","coinTypePair":"BTC","oid":"5aa499c60b9369642185a985","dealPrice":0.00125599,"orderOid":"5aa499c50b9369664b72606e","feeRate":0.001,"direction":"SELL"},{"coinType":"XRB","createdAt":1520735198000,"amount":0.95,"dealValue":0.00116771,"fee":9.5E-4,"dealDirection":"BUY","coinTypePair":"BTC","oid":"5aa493de0b9369642185a162","dealPrice":0.00122917,"orderOid":"5aa493dd0b936965a81c0d73","feeRate":0.001,"direction":"BUY"},"currPageNo":1,"limit":20,"pageNos":10}}')
[]

('\nRequest:', 'GET', 'https://api.kucoin.com/v1/order/detail?orderOid=5aa493dd0b936965a81c0d73&symbol=XRB-BTC&type=BUY', {'KC-API-SIGNATURE': 'xxx', 'KC-API-KEY': 'xxx', 'KC-API-NONCE': 'xxx', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36'}, None)
('\nResponse:', 'GET', 'https://api.kucoin.com/v1/order/detail?orderOid=5aa493dd0b936965a81c0d73&symbol=XRB-BTC&type=BUY', '200', "{'Access-Control-Allow-Methods': 'POST, GET, PUT, OPTIONS, DELETE, PATCH', 'KC-NODE': 'localhost', 'Access-Control-Max-Age': '86400', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Access-Control-Allow-Credentials': 'true', 'Date': 'Sun, 11 Mar 2018 12:18:08 GMT', 'Access-Control-Allow-Origin': 'https://www.kucoin.com', 'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept, Accept-Language', 'Content-Type': 'application/json;charset=UTF-8'}", u'{"success":true,"code":"OK","msg":"Operation succeeded.","timestamp":1520770688893,"data":{"coinType":"XRB","dealValueTotal":0.00116771,"feeTotal":9.5E-4,"userOid":"5a523ee6df542b7e58dc572b","dealAmount":0.95,"coinTypePair":"BTC","type":"BUY","orderOid":"5aa493dd0b936965a81c0d73","createdAt":1520735198000,"dealOrders":{"total":1,"firstPage":true,"lastPage":false,"datas":[{"createdAt":1520735198000,"amount":0.95,"dealValue":0.00116771,"fee":9.5E-4,"dealPrice":0.00122917,"feeRate":0.001}],"currPageNo":1,"limit":20,"pageNos":1},"dealPriceAverage":0.00122917,"orderPrice":0.00122917,"pendingAmount":0.0}}')
{'status': 'closed', 'timestamp': 1520735198000L, 'price': 0.00122917, 'datetime': '2018-03-11T02:26:38.000Z', 'cost': 0.00116771, 'id': '5aa493dd0b936965a81c0d73', 'remaining': 0.0, 'info': {u'dealOrders': {u'currPageNo': 1, u'firstPage': True, u'lastPage': False, u'limit': 20, u'pageNos': 1, u'total': 1, u'datas': [{u'fee': 0.00095, u'feeRate': 0.001, u'amount': 0.95, u'dealPrice': 0.00122917, u'dealValue': 0.00116771, u'createdAt': 1520735198000L}]}, u'orderPrice': 0.00122917, u'coinType': u'XRB', u'pendingAmount': 0.0, u'dealAmount': 0.95, u'dealPriceAverage': 0.00122917, u'feeTotal': 0.00095, u'coinTypePair': u'BTC', u'orderOid': u'5aa493dd0b936965a81c0d73', u'userOid': u'5a523ee6df542b7e58dc572b', u'dealValueTotal': 0.00116771, u'type': u'BUY', u'createdAt': 1520735198000L}, 'fee': {'currency': u'XRB', 'rate': None, 'cost': 0.00095}, 'trades': [{'timestamp': 1520735198000L, 'price': 0.00122917, 'datetime': '2018-03-11T02:26:38.000Z', 'cost': 0.00116771, 'id': None, 'info': {u'fee': 0.00095, u'feeRate': 0.001, u'amount': 0.95, u'dealPrice': 0.00122917, u'dealValue': 0.00116771, u'createdAt': 1520735198000L}, 'fee': {'currency': None, 'cost': 0.00095}, 'order': '5aa493dd0b936965a81c0d73', 'symbol': u'XRB/BTC', 'amount': 0.95, 'type': None, 'side': u'buy'}], 'side': u'buy', 'amount': 0.95, 'type': 'limit', 'symbol': u'XRB/BTC', 'filled': 0.95}
('\nRequest:', 'GET', 'https://api.kucoin.com/v1/order/detail?orderOid=5aa499c50b9369664b72606e&symbol=XRB-BTC&type=SELL', {'KC-API-SIGNATURE': 'xxx', 'KC-API-KEY': 'xxx', 'KC-API-NONCE': 'xxx', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36'}, None)
Traceback (most recent call last):
  File "C:/Users/Simon/Python/Binance/CCXT_Binance_Kucoin_playground.py", line 215, in <module>
    last_sell_order_kucoin = kucoin.fetch_order(id=sell_order_ID_kucoin, symbol='XRB/BTC', params={'type': "SELL"})
  File "C:\Users\Simon\Python\Binance\venv1\lib\site-packages\ccxt\kucoin.py", line 460, in fetch_order
    raise OrderNotFound(self.id + ' ' + self.json(response))
ccxt.base.errors.OrderNotFound: kucoin {"msg":"Operation succeeded.","timestamp":1520770690462,"code":"OK","data":null,"success":true}
('\nResponse:', 'GET', 'https://api.kucoin.com/v1/order/detail?orderOid=5aa499c50b9369664b72606e&symbol=XRB-BTC&type=SELL', '200', "{'Access-Control-Allow-Methods': 'POST, GET, PUT, OPTIONS, DELETE, PATCH', 'KC-NODE': 'localhost', 'Access-Control-Max-Age': '86400', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Access-Control-Allow-Credentials': 'true', 'Date': 'Sun, 11 Mar 2018 12:18:10 GMT', 'Access-Control-Allow-Origin': 'https://www.kucoin.com', 'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept, Accept-Language', 'Content-Type': 'application/json;charset=UTF-8'}", u'{"success":true,"code":"OK","msg":"Operation succeeded.","timestamp":1520770690462,"data":null}')

Process finished with exit code 1
```
Please note that I have omitted all the other orders produced by the _kucoin.fetch_closed_orders(symbol='XRB/BTC')_ to shorten the output pasted above.

Many thanks again in advance and I hope it's an easy fix

**Comment by gaardiolor** [comment]
@kroitor 
not too sure what happens, but the open_order info doesn't make it past `array = [entry for entry in array if (key in entry) and (entry[key] is not None)]` in `group_by()` in the exchange base class. 

The info going in that list comprehension:
```[('5aa414b90473f25f08ecd6d3',
  {'amount': 0.09305361000000001,
   'cost': 0.42571436351597997,
   'datetime': '2018-03-10T17:24:10.000Z',
   'fee': {'cost': None, 'currency': 'BCH', 'rate': None},
   'filled': 0.03769227,
   'id': '5aa414b90473f25f08ecd6d3',
   'info': {'coinType': 'BCH',
            'coinTypePair': 'NEO',
            'createdAt': 1520702650000,
            'dealAmount': 0.03769227,
            'direction': 'BUY',
            'oid': '5aa414b90473f25f08ecd6d3',
            'pendingAmount': 0.05536134,
            'price': 11.294474,
            'status': 'open',
            'updatedAt': 1520702650000,
            'userOid': None},
   'price': 11.294474,
   'remaining': 0.05536134,
   'side': 'buy',
   'status': 'open',
   'symbol': 'BCH/NEO',
   'timestamp': 1520702650000,
   'trades': None,
   'type': 'limit'})]```

So the `entry` is ```
('5aa414b90473f25f08ecd6d3',
  {'amount': 0.09305361000000001,
   'cost': 0.42571436351597997,
   'datetime': '2018-03-10T17:24:10.000Z',
   'fee': {'cost': None, 'currency': 'BCH', 'rate': None},
   'filled': 0.03769227,
   'id': '5aa414b90473f25f08ecd6d3',
   'info': {'coinType': 'BCH',
            'coinTypePair': 'NEO',
            'createdAt': 1520702650000,
            'dealAmount': 0.03769227,
            'direction': 'BUY',
            'oid': '5aa414b90473f25f08ecd6d3',
            'pendingAmount': 0.05536134,
            'price': 11.294474,
            'status': 'open',
            'updatedAt': 1520702650000,
            'userOid': None},
   'price': 11.294474,
   'remaining': 0.05536134,
   'side': 'buy',
   'status': 'open',
   'symbol': 'BCH/NEO',
   'timestamp': 1520702650000,
   'trades': None,
   'type': 'limit'})``` 

instead of
```
{'amount': 0.09305361000000001,
   'cost': 0.42571436351597997,
   'datetime': '2018-03-10T17:24:10.000Z',
   'fee': {'cost': None, 'currency': 'BCH', 'rate': None},
   'filled': 0.03769227,
   'id': '5aa414b90473f25f08ecd6d3',
   'info': {'coinType': 'BCH',
            'coinTypePair': 'NEO',
            'createdAt': 1520702650000,
            'dealAmount': 0.03769227,
            'direction': 'BUY',
            'oid': '5aa414b90473f25f08ecd6d3',
            'pendingAmount': 0.05536134,
            'price': 11.294474,
            'status': 'open',
            'updatedAt': 1520702650000,
            'userOid': None},
   'price': 11.294474,
   'remaining': 0.05536134,
   'side': 'buy',
   'status': 'open',
   'symbol': 'BCH/NEO',
   'timestamp': 1520702650000,
   'trades': None,
   'type': 'limit'}
```

**Comment by kroitor** [comment]
> not too sure what happens, but the open_order info doesn't make it past array = [entry for entry in array if (key in entry) and (entry[key] is not None)] in group_by() in the exchange base class.

Sorry, but this description does not really help without knowing the `key` value (the field name) from there... It would be much easier for me if you could paste code + verbose output as described here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-submit-an-issue, I would then be able to fix it quicker.

**Comment by kroitor** [comment]
@gaardiolor thx anyway, no worries, will fix it soon )

**Comment by gaardiolor** [comment]
Yeah, slix88 already pasted the verbose stuff. 

Not sure if it is wat you want (probably not :) ), but this works for me:
```
    async fetchOpenOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
        if (!symbol)
            throw new ExchangeError (this.id + ' fetchOpenOrders requires a symbol');
        await this.loadMarkets ();
        let market = this.market (symbol);
        let request = {
            'symbol': market['id'],
        };
        let response = await this.privateGetOrderActiveMap (this.extend (request, params));
        let orders = this.arrayConcat (response['data']['SELL'], response['data']['BUY']);
        let result = [];
        for (let i = 0; i < orders.length; i++) {
            let order = this.parseOrder (this.extend (orders[i], {
                'status': 'open',
            }), market);
            result.push (order);
        }
        return result;
    }
```

**Comment by kroitor** [comment]
> First one is that "fetch_closed_orders" is no longer producing an error (which is the fix above). > However, it's not returning anything.

This particular issue is now fixed in ccxt 1.11.81, so, `fetch_closed_orders` will return an adequate response now.

Working on through remaining issues here.

**Comment by gaardiolor** [comment]
ah, the original issue was about fetch_closed_orders, sorry, missed that :( Hopefully it's related, otherwise I'd feel guilty for not creating a new issue!
